File: apps/api/app/schemas/honor_system.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class RankProfileRead(BaseModel):
    """
    A user's aggregate rank state in a single (sport, area).

    ``id``, ``created_at``, and ``updated_at`` are nullable to support
    the non-persisted default returned by ``GET /rankings/me`` for a
    brand-new user — that endpoint is read-only and must not create a
    row, so the response has no DB-assigned identifiers to surface. A
    persisted profile always has all three populated.
    """

    id: UUID | None
    user_id: UUID
    sport: str
    area: str
    rating: int
    wins: int
    losses: int
    streak: int
    last_played_at: datetime | None
    created_at: datetime | None
    updated_at: datetime | None

    model_config = {"from_attributes": True}


class RankingEntry(BaseModel):
    """One row of a leaderboard. ``rank`` is 1-based, dense."""

    rank: int
    user_id: UUID
    rating: int
    wins: int
    losses: int
    streak: int


class RankingListResponse(BaseModel):
    sport: str
    area: str
    items: list[RankingEntry]
    total: int


class HonorTitleRead(BaseModel):
    id: UUID
    sport: str
    area: str
    title_name: str
    current_holder_user_id: UUID | None
    active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class HonorHistoryRead(BaseModel):
    id: UUID
    honor_title_id: UUID
    previous_holder_user_id: UUID | None
    new_holder_user_id: UUID
    source_match_id: UUID | None
    created_at: datetime

    model_config = {"from_attributes": True}


class HonorTransferResponse(BaseModel):
    """
    Return type of
    :func:`app.services.honor_system.record_match_result_for_honor`.

    Includes both rank-profile updates so the caller can refresh local
    state in one round-trip, the current honor title state, and a
    ``transferred`` flag indicating whether the title moved in this
    call. ``history_entry`` is populated only when a transfer occurred
    (initial award or holder swap).

    Used by the service layer and by the future verified-result hook.
    Intentionally not exposed via a public POST today — see
    :mod:`app.routers.honor_system` for the rationale.
    """

    winner_profile: RankProfileRead
    loser_profile: RankProfileRead
    honor_title: HonorTitleRead
    transferred: bool
    history_entry: HonorHistoryRead | None

File: packages/shared-types/src/honor-system.ts
/**
 * Honor System (local champion titles) — read-only public API contract.
 *
 * Mirrors the backend schemas in ``apps/api/app/schemas/honor_system.py``.
 * The mobile API client auto-transforms response keys from snake_case to
 * camelCase, so the TypeScript shapes below use camelCase. Date fields
 * are ISO 8601 strings; UUID fields are string-typed UUIDs.
 *
 * Distinct from {@link ./rank.ts}, which models the legacy event-driven
 * ``/rank/me`` Honor / Gang Score summary. This module models the
 * per-(sport, area) leaderboard + champion-title surfaces used by
 * Profile's LocalRankSection.
 *
 * READ-ONLY by design — there are no mutation request types here. The
 * single backend writer is the verified challenge / tournament /
 * group-event result hook; it is not exposed to the public mobile API.
 */

import type { ISODateString, UUID } from './common';

/**
 * A user's aggregate rank state in a single (sport, area).
 *
 * ``id``, ``createdAt`` and ``updatedAt`` are nullable to support the
 * non-persisted default returned by ``GET /rankings/me`` for a
 * brand-new user — that endpoint is read-only and never inserts, so
 * the response has no DB-assigned identifiers to surface in that
 * shape. A persisted profile always has all three populated.
 *
 * ``sport`` and ``area`` are stored as freeform strings on the backend
 * (``String(30)`` / ``String(80)``) and are NOT restricted to the
 * ``Sport`` literal — the Honor System intentionally supports a wider
 * sport vocabulary than venues.
 */
export interface RankProfileRead {
  id: UUID | null;
  userId: UUID;
  sport: string;
  area: string;
  rating: number;
  wins: number;
  losses: number;
  streak: number;
  lastPlayedAt: ISODateString | null;
  createdAt: ISODateString | null;
  updatedAt: ISODateString | null;
}

/** One row of an (sport, area) leaderboard. ``rank`` is 1-based, dense. */
export interface RankingEntry {
  rank: number;
  userId: UUID;
  rating: number;
  wins: number;
  losses: number;
  streak: number;
}

export interface RankingListResponse {
  sport: string;
  area: string;
  items: RankingEntry[];
  total: number;
}

/**
 * The "{Area} {Sport} Champion" honor title for one (sport, area).
 * ``currentHolderUserId`` is null only briefly between creation and
 * the first holder assignment — in practice this surface always sees
 * a populated holder.
 */
export interface HonorTitleRead {
  id: UUID;
  sport: string;
  area: string;
  titleName: string;
  currentHolderUserId: UUID | null;
  active: boolean;
  createdAt: ISODateString;
  updatedAt: ISODateString;
}
